### PR TITLE
Support Future for collection filters

### LINKED ISSUE #722: NH-1953 - Support Future for collection filters
**Marko Podgorsek** created an [issue](http://nhibernate.jira.com/browse/NH-1953) — <time datetime="2009-09-02T06:57:10.844Z" title="2009-09-02T06:57:10.844Z">2nd September 2009, 6:57:10</time>:
> Currently, `Future<>` doesn't work on `CreateFilter`
>
> I have a class:
> 
> ```csharp
> public class User
> {
>       public virtual IList<Tag> Tags { get; protected set; }
> }
> ```
> 
> 
> public class Tag .... //not important
> 
> Mapping for User -- the Tags part:
> 
> ```xml
> <bag lazy="true" name="Tags" table="USER_TAG">
>       <key>
>         <column name="ID_USER" />
>       </key>
>       <many-to-many class="BeriMe.Biz.Domain.Tag, BeriMe.Biz, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null">
>         <column name="ID_TAG" />
>       </many-to-many>
> </bag>
> ```
> 
> 
> If I do this:
> 
> ```csharp
> User u = db.Get<User>(5);
> 
> var t = db.CreateFilter(u.Tags, "")
>                 .SetFirstResult(4)
>                 .SetMaxResults(3)
>                 .Future<Tag>();    //If I say List<Tag>() it works OK.
> 
>             foreach (var x in t) { } //crash
> 
> ```
> 
> Error:
> NHibernate.Hql.Ast.ANTLR.QuerySyntaxException: Exception of type 'Antlr.Runtime.MismatchedTreeNodeException' was thrown. [].
> 
> Error from log:
> ERROR NHibernate.Hql.Parser - MismatchedTreeNodeException(3!=2)
> 
> 
> 
---
**Fabio Maulo** added a comment — <time datetime="2009-09-02T23:20:15.062Z" title="2009-09-02T23:20:15.062Z">2nd September 2009, 23:20:15</time>:
> I think that was never supported.
>

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #722

### CODE REVIEW

**Comment by hazzik** [comment]
Need to override ToList in LinqFilterImpl

**Comment by bahusoid** [comment]
Oh.. Thanks. Will fix tommorow. But why firebird only?

**Comment by hazzik** [comment]
Because it does not support multi-query and that code path executes `List<T>` instead of `List`.

File: src/NHibernate/Impl/ExpressionQueryImpl.cs [hunk 1]
```diff
@@ -125,6 +125,32 @@ public override IList List()
 			}
 		}
 
+		public override IList<T> List<T>()
```
Comment (hazzik) [inline src/NHibernate/Impl/ExpressionQueryImpl.cs]: Lack of this override caused to call `Session.List` and this resulted in a lost `collection` parameter. So the query applied to the whole table.

Comment (fredericDelaporte) [reply src/NHibernate/Impl/ExpressionQueryImpl.cs]: For correctness of the class, shouldn't a `void List(IList results)` override be added too? (And same for `CollectionFilterImpl`.)

Comment (bahusoid) [reply src/NHibernate/Impl/ExpressionQueryImpl.cs]: Done. I've also added missing Enumerable ovverrides to `ExpressionQueryImpl` to throw NotSupportedException.

**Pull request merged**